### PR TITLE
add email verification

### PR DESCRIPTION
**Pull request by M437A**
add:
- abstract class for email sending
- abstract message builder
- add email verification 

how it works:

Register
- Send post request about registrar a new user
- Generate temporary token
- Save data from this request to the cache as value and token as a key
(also we have another hash-table, were we check, that we can send just one code in 5 min)

Confirm 
- Get data by token
- Check connection between email and token
- Save user data to db

### CODE REVIEW

File: src/main/java/com/zufar/icedlatte/email/sender/AbstractEmailSander.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.mail.SimpleMailMessage;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Locale;
+
+@Service
+@RequiredArgsConstructor
+public abstract class AbstractEmailSander<T> {
```
Comment (yevr19) [inline src/main/java/com/zufar/icedlatte/email/sender/AbstractEmailSander.java:14]: typo

File: src/main/java/com/zufar/icedlatte/email/sender/AbstractEmailSander.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        mailMessage.setText(message);
+        mailMessage.setSubject(subject);
+        javaMailSender.send(mailMessage);
+    }
+
+    protected String getMessage(Class<?> clazz, T event) {
+        return messageBuilders.stream()
+                .filter(builder -> builder.supports(clazz))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("No message builder found"))
```
Comment (yevr19) [inline src/main/java/com/zufar/icedlatte/email/sender/AbstractEmailSander.java:31]: can we use some specific runtime exception here?

File: src/main/java/com/zufar/icedlatte/email/sender/EmailConfirmation.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Autowired
+    public EmailConfirmation(JavaMailSender javaMailSender,
+                             SimpleMailMessage mailMessage,
+                             List<MessageBuilder<EmailConfirmationDto>> messageBuilders) {
+        super(javaMailSender, mailMessage, messageBuilders);
+    }
+
+    public void sendTemporaryCode(String email, String message) {
+        message = getMessage(EmailConfirmMessage.class, new EmailConfirmationDto(message));
```
Comment (yevr19) [inline src/main/java/com/zufar/icedlatte/email/sender/EmailConfirmation.java:28]: create a new variable here as it can cause side effects

**Review by yevr19 — APPROVED** [review]

**Review by Sunagatov — APPROVED** [review]
Approved

**Pull request merged**